### PR TITLE
ENH: natively support Graph in all Join Counts statistics

### PR DESCRIPTION
**Pull request by martinfleis**
Adding a native support to all Join Counts classes.

This uncovered two bugs in libpysal, one in `eliminate_zeros` and other in `isolates`. I have fixes for both locally, a PR is coming. Marking this as a draft until these land in lib.

### CODE REVIEW

**Review by jGaboardi — APPROVED** [review]

File: esda/join_counts.py [hunk 1]
```diff
@@ -150,12 +152,16 @@ def __init__(self, y, w, permutations=PERMUTATIONS, drop_islands=True):
...
+        if isinstance(w, W):
+            self.adj_list = self.w.to_adjlist(
+                remove_symmetric=False, drop_islands=drop_islands
+            )
+        else:
+            self.adj_list = self.w.adjacency.reset_index()
         self.y = y
         self.permutations = permutations
-        self.J = w.s0 / 2.0
+        s0 = w._s0 if hasattr(w, "_s0") else w.s0
```
Comment (sjsrey) [inline esda/join_counts.py:163]: I'm not clear on why these need to be private (s0 in this case)? What is the advantage of that?

Comment (martinfleis) [reply esda/join_counts.py:163]: @ljwolf didn't like exposing them as public properties. I originally mirrored weights API which was easy and I'm happy to follow whichever decision we make in lib.

**Comment by martinfleis** [comment]
Dev is green 🎉

**Review by knaaptime — APPROVED** [review]

**Comment by jGaboardi** [comment]
* CI is all green - `libpysal` is pinned - we have approvals
* I'll merge tomorrow in the AM unless someone else already has or there are detractors

**Pull request merged**